### ISSUE TITLE
`quantile` calculation using dask default method gives incorrect results

### ISSUE THREAD

**Issue by r-luo** [dask/dask#6551]
This is similar to #731 but reveals a broader issue and thus I would like to create a new issue for it. I'm using `dask==2.24.0` and `pandas==1.0.5`
```
import pandas as pd
import dask.dataframe as dd
x = pd.DataFrame({'a': [1,2,3]})
xd = dd.from_pandas(x, npartitions=3)
print(x['a'].quantile(0.5))
print(xd['a'].quantile(0.5).compute())
```

This difference persists as I try all natural number column `'a'` with length greater than 2 (i.e. `'a'` being natural numbers 1,...,N where N>2).

```
for i in range(1, 100):
    x = pd.DataFrame({'a': list(range(i))})
    xd = dd.from_pandas(x, npartitions=i+1)
    pd_quantile = x['a'].quantile(0.5)
    dd_quantile = xd['a'].quantile(0.5).compute()
    if pd_quantile == dd_quantile:
        print('same')
    else:
        print('different')
```

Using `method='tdigest'` does solve the issue, but the default algorithm shouldn't be giving a wrong answer either..

**Comment by TomAugspurger** [comment]
How is this different than #731?

**Comment by r-luo** [comment]
@TomAugspurger It's different in the sense that this error actually happens for simple cases as opposed to just the edge cases described in #731

**Comment by TomAugspurger** [comment]
Thanks. Do you think that fixing #731 is likely to fix this as well?

And I haven't looked closely at the root cause, but does having as many partitions as points make this problem likely to occur? I think that'd be uncommon in practice.

**Comment by r-luo** [comment]
@TomAugspurger I think this is probably the same issue, as I just did another test with random numbers. The conclusion is:
- the quantile calculation in dask dataframe is only correct when `npartitions=1`
- the average absolute error of dask's quantile calculation declines as the ratio of (number of data points / number of partitions) increases, i.e. either when data size increases with fixed partition size, or when number of partition decreases for fixed amount of data

I think we can conclude that there must be something wrong with the reduce step for the quantile calculation.

Here is the code I used to test it:
```python
import pandas as pd
import numpy as np
import dask.dataframe as dd

# for fixed npartitions, see the effect of increasing dataset size

data_sizes = [2 * 10**i for i in range(5)]
npartitions = 2
size_res = {}
for s in data_sizes:
    # 100 experiments for each data size
    for j in range(100):
        size_res[s] = []
        x = pd.DataFrame({'a': np.random.uniform(size=s)})
        xd = dd.from_pandas(x, npartitions=npartitions)
        xm = x['a'].quantile()
        xdm = xd['a'].quantile().compute()
        size_res[s].append(xm-xdm)

# Calculate the mean absolute error of dask's quantile calculation at each dataset size
size_mae = {}
for s in size_res:
    size_mae[s] = np.mean(np.abs(size_res[s]))

#size_mae={
#    2: 0.0,
#    20: 0.023239827911832822,
#    200: 0.011693968030884405,
#    2000: 0.030041772340274497,
#    20000: 0.0055823681147273985
#}

# For fixed data size, see the effect of increasing partitions

data_size = 10000
npartitions = [4**i for i in range(7)]
npar_res = {}
for npar in npartitions:
    # 100 experiments for each partition size
    for j in range(100):
        npar_res[npar] = []
        x = pd.DataFrame({'a': np.random.uniform(size=data_size)})
        xd = dd.from_pandas(x, npartitions=npar)
        xm = x['a'].quantile()
        xdm = xd['a'].quantile().compute()
        npar_res[npar].append(xm-xdm)


# Calculate the mean absolute error of dask's quantile calculation at each npartitions
npar_mae = {}
for s in npar_res:
    npar_mae[s] = np.mean(np.abs(npar_res[s]))

# npar_mae={
#    1: 0.0,
#    4: 0.00337763961928661,
#    16: 0.02724938859132664,
#    64: 0.11335921282273076,
#    256: 0.21650419213080002,
#    1024: 0.3861034689856373,
#    4096: 0.4905171995922507
#}

```

**Comment by r-luo** [comment]
Even further interesting results - The ordering of the original array matters:

```python
res = {}
base = np.array([0,0,0,0,0,1,1])
cur = list(base.copy())
while (not (np.array(cur) == base).all() or len(res) == 0):
    x = pd.DataFrame({'a': cur})
    xd = dd.from_pandas(x, npartitions=3)
    xm = x['a'].quantile()
    xdm = xd['a'].quantile().compute()
    res[','.join([str(j) for j in cur])] = xm - xdm
    cur.insert(0, cur.pop())

```

![image](https://user-images.githubusercontent.com/3460073/91253129-234cc600-e72d-11ea-9169-fda54a1b3735.png)

**Comment by eriknw** [comment]
Neat.  I'm curious, how do things behave if we always include 0 and 1 in the quantiles to calculate?
```python
In [17]: xd['a'].quantile([0.5]).compute()
Out[17]:
0.5    2.5
Name: a, dtype: float64

In [18]: xd['a'].quantile([0, 0.5, 1]).compute()
Out[18]:
0.0    1.0
0.5    2.0
1.0    3.0
Name: a, dtype: float64
```
Adding 0 and 1 internally (if necessary) could be an easy fix for edge cases such as these, and should generally make quantile calculations better.

**Comment by r-luo** [comment]
@eriknw That's very interesting - why does adding 0 and 1 in quantile calculation change the algorithm's behaviour?

**Comment by r-luo** [comment]
More interesting facts:

As long as you add anything greater than 0.5, the quantile calculation at 0.5 will be correct:
```python
res = {}
base = np.array([0,0,0,0,0,1,1])
cur = list(base.copy())
while (not (np.array(cur) == base).all() or len(res) == 0):
    x = pd.DataFrame({'a': cur})
    xd = dd.from_pandas(x, npartitions=3)
    xm = x['a'].quantile([0.5, 0.50000000001])
    xdm = xd['a'].quantile([0.5, 0.50000000001]).compute()
    res[','.join([str(j) for j in cur])] = xm - xdm
    cur.insert(0, cur.pop())
print(res)
```
![image](https://user-images.githubusercontent.com/3460073/91450912-d65f1180-e84a-11ea-9234-f6d18e52bb9a.png)

**Comment by r-luo** [comment]
@eriknw I just found that adding 0 and 1 quantiles don't fully fix the problem:

```python
res = {}
base = np.array([0,0,0,0,0,1,1])
cur = list(base.copy())
while (not (np.array(cur) == base).all() or len(res) == 0):
    x = pd.DataFrame({'a': cur})
    xd = dd.from_pandas(x, npartitions=3)
    xm = x['a'].quantile([0, 0.4, 0.5, 0.6, 1])
    xdm = xd['a'].quantile([0, 0.4, 0.5, 0.6, 1]).compute()
    res[','.join([str(j) for j in cur])] = xm - xdm
    cur.insert(0, cur.pop())
print(res)
```

![image](https://user-images.githubusercontent.com/3460073/91451452-7026be80-e84b-11ea-9d8e-2056de39e6cc.png)

**Comment by jcrist** [comment]
This isn't 100% fixed (the real fix is to use tdigest everywhere), but with #6564 things are a bit better. I consider the actual fix to be done by resolving #6566, closing this issue to continue tracking there.
